Add funcorder linter

manuelarte opened the pull request:
This linter detects that the exported method for a struct are placed before the non-exported methods. It also detects that the `NewXYZ` functions are placed after the struct declaration and before the struct's methods.
Following the uber style guidelines about [function-grouping-and-ordering](https://github.com/uber-go/guide/blob/master/style.md#function-grouping-and-ordering).

Repository: https://github.com/manuelarte/funcorder

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/golangci/golangci-lint?pullRequest=5630) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/golangci/golangci-lint?pullRequest=5630) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: manuelarte<br/>:x: Manuel Doncel Martos<br/><hr/>**Manuel Doncel Martos** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/golangci/golangci-lint?pullRequest=5630) it.</sub>

manuelarte commented:
For some reason, when I try to run:

>  go run ./cmd/golangci-lint/ run --no-config --default=none --enable=gofuncor ./pkg/golinters/gofuncor/testdata/gofuncor.go

I get the errors that I am expecting, but the `// want ...` comment is not "working" (even that I have the same testdata in my repo and it's working)/

ldez commented:
/
In order for a pull request adding a linter to be reviewed, the linter and the PR must follow some requirements.

- [ ] The CLA must be signed

### Pull Request Description

- [ ] It must have a link to the linter repository.
- [ ] It must provide a short description of the linter.

### Linter

- [ ] It must not be a duplicate of another linter or a rule of a linter (the team will help to verify that).
- [ ] It must have a valid license (AGPL is not allowed), and the file must contain the required information by the license, ex: author, year, etc.
- [ ] It must use Go version <= 1.23
- [ ] The linter repository must have a CI and tests.
- [ ] It must use [`go/analysis`](https://golangci-lint.run/contributing/new-linters/).
- [ ] It must have a valid tag, ex: `v1.0.0`, `v0.1.0`.
- [ ] It must not contain `init()`.
- [ ] It must not contain `panic()`.
- [ ] It must not contain `log.Fatal()`, `os.Exit()`, or similar.
- [ ] It must not modify the AST.
- [ ] It must not have false positives/negatives (the team will help to verify that).
- [ ] It must have tests inside golangci-lint.

### The Linter Tests Inside Golangci-lint

- [ ] They must have at least one std lib import.
- [ ] They must have integration tests without configuration (default).
- [ ] They must have integration tests with configuration (if the linter has a configuration).

### `.golangci.next.reference.yml`

- [ ] The file `.golangci.next.reference.yml` must be updated.
- [ ] The file `.golangci.reference.yml` must NOT be edited.
- [ ] The linter must be added to the lists of available linters (alphabetical case-insensitive order).
    - `enable` and `disable` options
- [ ] If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
    - The values must be different from the default ones.
    - The default values must be defined in a comment.
    - The option must have a short description.

### Others Requirements

- [ ] The files (tests and linter) inside golangci-lint must have the same name as the linter.
- [ ] The `.golangci.yml` of golangci-lint itself must not be edited and the linter must not be added to this file.
- [ ] The linters must be sorted in the alphabetical order (case-insensitive) in the `lintersdb/builder_linter.go` and `.golangci.next.reference.yml`.
- [ ] The load mode (`WithLoadMode(...)`):
    - if the linter uses `goanalysis.LoadModeSyntax` -> no `WithLoadForGoAnalysis()` in `lintersdb/builder_linter.go`
    - if the linter uses `goanalysis.LoadModeTypesInfo`, it requires `WithLoadForGoAnalysis()` in `lintersdb/builder_linter.go`
- [ ] The version in `WithSince(...)` must be the next minor version (`v1.X.0`) of golangci-lint.
- [ ] `WithURL()` must contain the URL of the repository.

### Recommendations

- [ ] The file `jsonschema/golangci.next.jsonschema.json` should be updated.
- [ ] The file `jsonschema/golangci.jsonschema.json` must NOT be edited.
- [ ] The linter repository should have a readme and linting.
- [ ] The linter should be published as a binary (useful to diagnose bug origins).
- [ ] The linter repository should have a `.gitignore` (IDE files, binaries, OS files, etc. should not be committed)
- [ ] A tag should never be recreated.
- [ ] use `main` as the default branch name.

---

The golangci-lint team will edit this comment to check the boxes before and during the review.

The code review will start after the completion of those checkboxes (except for the specific items that the team will help to verify).

The reviews should be addressed as commits (no squash).

If the author of the PR is a member of the golangci-lint team, he should not edit this message.

**This checklist does not imply that we will accept the linter.**

manuelarte commented:
Hi @ldez, I am checking the list that the PR needs to have, and I already "fixed" these two:
- The file .golangci.next.reference.yml must be updated.
- The version in WithSince(...) must be the next minor version (v2.X.0) of golangci-lint.

But I don't know if I need to continue fixing them, or should I wait until you "finish" the review?

Thanks for your time!

ldez commented on pkg/config/formatters_settings.go:
```diff
@@ -19,6 +19,7 @@ type FormatterSettings struct {
 	Gci       GciSettings       `mapstructure:"gci"`
 	GoFmt     GoFmtSettings     `mapstructure:"gofmt"`
 	GoFumpt   GoFumptSettings   `mapstructure:"gofumpt"`
+	GoFuncOr  GoFuncOrSettings  `mapstructure:"gofuncor"`
```
```suggestion
```

manuelarte replied:
applied

ldez commented on pkg/config/formatters_settings.go:
```diff
@@ -48,6 +49,8 @@ type GoFumptSettings struct {
 	LangVersion string `mapstructure:"-"`
 }
 
+type GoFuncOrSettings struct{}
+
```
```suggestion
```

manuelarte replied:
applied

ldez commented on pkg/golinters/gofuncor/gofuncor.go:
```diff
@@ -0,0 +1,19 @@
+	"golang.org/x/tools/go/analysis"
+)
+
+func New(_ *config.GoFuncOrSettings) *goanalysis.Linter {
```
```suggestion
func New() *goanalysis.Linter {
```

ldez commented on pkg/lint/lintersdb/builder_linter.go:
```diff
@@ -335,6 +336,11 @@ func (LinterBuilder) Build(cfg *config.Config) ([]*linter.Config, error) {
 			WithAutoFix().
 			WithURL("https://github.com/mvdan/gofumpt"),
 
+		linter.NewConfig(gofuncor.New(&cfg.Linters.Settings.GoFuncOr)).
```
```suggestion
		linter.NewConfig(gofuncor.New()).
```

ldez commented on pkg/lint/lintersdb/builder_linter.go:
```diff
@@ -335,6 +336,11 @@ func (LinterBuilder) Build(cfg *config.Config) ([]*linter.Config, error) {
 
+		linter.NewConfig(gofuncor.New(&cfg.Linters.Settings.GoFuncOr)).
+			WithSince("v2.1.0").
+			WithAutoFix().
```
```suggestion
```

manuelarte replied:
applied

manuelarte commented:
Fixed merge conflicts

alexandear commented on pkg/golinters/gofuncor/testdata/gofuncor.go:
Please add the file `pkg/golinters/gofuncor/gofuncor_integration_test.go` with tests. According to the checklist item, "They must have integration tests without configuration (default)."

See https://github.com/golangci/golangci-lint/blob/47e23e588a2ecde79d49588a7cb0acfbc8858b1e/pkg/golinters/exptostd/exptostd_integration_test.go#L9-L11 as an example

manuelarte replied:
added

alexandear commented on pkg/golinters/gofuncor/testdata/gofuncor.go:
```diff
@@ -0,0 +1,39 @@
+package testdata
```
Tests won't work without this line:
```suggestion
//golangcitest:args -Egofuncor
package testdata
```

manuelarte replied:
line added and now the integration test works

manuelarte commented on pkg/golinters/gofuncor/gofuncor_integration_test.go:
```diff
@@ -0,0 +1,10 @@
+
+import (
+	"github.com/golangci/golangci-lint/v2/test/testshared/integration"
+	"testing"
```
done!

manuelarte commented on pkg/golinters/gofuncor/testdata/gofuncor.go:
```diff
@@ -0,0 +1,40 @@
+	m.Name = name
+}
+
+func NewMyStruct2() *MyStruct2 { // want `constructor \"NewMyStruct2\" for struct \"MyStruct2\" should be placed before struct method \"GetName\"`
```
done

ldez commented:
For reference, this linter is close to #4937 (but it's different).

I need to think about the scope and its usage.

Note: the linter panics, the examples inside the linter readme contain typos, and the code style is more Java style than Go style.
I'm also not sure about the name: the prefix `go` seems useless, and a linter name should try to be understandable at first read, IMHO it's currently not the case because `goFuncOr` seems related to a logical `OR` and not to `order.

alexandear commented:
Perhaps the linter is similar to [decorder](https://golangci-lint.run/usage/linters/#decorder)

ldez commented:
Not similar, but complementary.

They handle different topics.

manuelarte commented:
> For reference, this linter is close to #4937 (but it's different).
> 
> I need to think about the scope and its usage.
> 
> Note:
> 
> The linter panics.
> 
> ```
> ERRO [runner] Panic: gofuncor: package "internal" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference: goroutine 1101 [running]:
> runtime/debug.Stack()
>         /usr/lib/go/src/runtime/debug/stack.go:26 +0x5e
> github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyzeSafe.func1()
>         /home/ldez/sources/golangci/golangci-lint/pkg/goanalysis/runner_action.go:50 +0x257
> panic({0x1b708e0?, 0x2ae34e0?})
>         /usr/lib/go/src/runtime/panic.go:792 +0x132
> go/ast.(*FuncDecl).Pos(0xc00389f9a0?)
>         /usr/lib/go/src/go/ast/ast.go:1007 +0xe
> github.com/manuelarte/gofuncor/internal/models.(*StructHolder).Analyze(0xc003e80340)
>         /home/ldez/sources/go/pkg/mod/github.com/manuelarte/gofuncor@v0.1.0/internal/models/struct_holder.go:64 +0x507
> github.com/manuelarte/gofuncor/internal/fileprocessor.(*FileProcessor).Analyze(0x206c7e0?)
>         /home/ldez/sources/go/pkg/mod/github.com/manuelarte/gofuncor@v0.1.0/internal/fileprocessor/file_processor.go:48 +0xb1
> github.com/manuelarte/gofuncor/pkg/analyzer.run(0xc001ad3a40)
>         /home/ldez/sources/go/pkg/mod/github.com/manuelarte/gofuncor@v0.1.0/pkg/analyzer/analyzer.go:39 +0x16a
> github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*action).analyze.func3(...)
>         /home/ldez/sources/golangci/golangci-lint/pkg/goanalysis/runner_checker.go:182
> ```

Can I know what command did you run to make it panic?


> The examples inside the linter readme contain typos, and so don't compile (because of the typos).
> 

I think I fixed them.

> The code style is more Java style (ex: `PrivateMethodBeforePublicForStructTypeError`, `ConstructorNotBeforeStructMethodsError`, etc.) than Go style (this is not a problem, just a note).
> 

I am open to suggestions on improvement, I have a Java background and maybe that's why I have that mentality.

> I'm also not sure about the name: the prefix `go` seems useless, and a linter name should try to be understandable at first read, IMHO it's currently not the case because `goFuncOr` seems related to a logical `OR` and not to `order`.

I am also open to suggestions here, maybe GoFuncOrd, or FuncOrd, or something related to grouping and ordering like FuncGroupOrd.

ldez commented:
> Can I know what command (or what testdata did you use) did you run to make it panic?

```bash
./golangci-lint run --enable-only gofuncor
```

On the golangci-lint code.

manuelarte commented:
> > Can I know what command (or what testdata did you use) did you run to make it panic?
> 
> ```shell
> ./golangci-lint run --enable-only gofuncor
> ```
> 
> On the golangci-lint code.

checking it

ldez commented:
> I am also open to suggestions here

I suggest something simple like `funcOrder` or `funcSort`.

> I have a Java background and maybe that's why I have that mentality.

Note that no one uses the word "Golang" outside of a search term :wink:

manuelarte commented:
@ldez I think I fixed the panic, and when I run 

> ./golangci-lint run --enable-only gofuncor

I can see my linter in action.

Regarding the name, I like `funcOrder`, what do you think it's the best approach?

- new repo from scratch
- rename current repo and all the references
?

ldez commented:
I think renaming the repo (and the linter) is the best approach

manuelarte commented:
> I think renaming the repo (and the linter) is the best approach

Ok, I think I did it. I renamed to `funcorder`, and made the changes in the repo. Hope everything looks fine now.

manuelarte commented:
@ldez @alexandear I added you both as (I think it's called), collaborator of funcorder, feel free to accept or reject the invitation.

ldez commented:
Before, could you explain why?

manuelarte commented:
> Before, could you explain why?

I was talking to @ccoVeille and he recommended me that it was a good idea, but if you don't agree, like it, etc, I can remove it.

ldez commented:
I don't understand why @ccoVeille suggested that.

Anyway, I will help you.

ccoVeille commented:
Wow, such an exposure... 

Got direct messages from everyone.

I hope I sorted the misunderstandings.

Take care

ldez commented on pkg/golinters/funcorder/funcorder.go:
```diff
@@ -1,19 +1,39 @@
 	"golang.org/x/tools/go/analysis"
 
+	"github.com/golangci/golangci-lint/v2/pkg/config"
+
```
```suggestion
```

manuelarte replied:
solved

manuelarte commented:
I updated this PR with the new version of `funcorder` with the settings feature.

The part of the settings looks like this:

```go
func New(settings *config.FuncOrderSettings) *goanalysis.Linter {
	a := analyzer.NewAnalyzer()

	cfg := map[string]map[string]any{}

	if settings != nil {
		constructor := true
		if settings.Constructor != nil {
			constructor = *settings.Constructor
		}
		structMethod := true
		if settings.StructMethod != nil {
			structMethod = *settings.StructMethod
		}

		cfg[a.Name] = map[string]any{
			"constructor":   constructor,
			"struct-method": structMethod,
		}
	}

	return goanalysis.NewLinter(
		a.Name,
		a.Doc,
		[]*analysis.Analyzer{a},
		cfg,
	).WithLoadMode(goanalysis.LoadModeSyntax)
}
```

As you can see I needed to "hardcode" `constructor` and `struct-method`, because those two constants are not exported in `funcorder` repo, does it make sense to export them and then use them like:

```go
cfg[a.Name] = map[string]any{
    analyzer.Constructor:   constructor,
    analyzer.SstructMethod: structMethod,
}
```
?

----

I also updated `.golangci.next.reference` with the settings and an explanation.

ldez commented:
Yes, you can export the constants.

manuelarte commented:
I fixed the check:

If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
- The values must be different from the default ones.
- The default values must be defined in a comment.
- The option must have a short description.

By setting the values to false and adding the comment
```
# Default: true
```

manuelarte commented:
is there something else expected from me? or I just need to wait for this PR to be approved or rejected?

ldez commented:
For now, you just have to wait.

alexandear commented on .golangci.next.reference.yml:
```diff
@@ -529,6 +531,14 @@ linters:
       analyze-types: true
 
+    funcorder:
+      # Enable/disable feature to check constructors are placed after struct declaration
```
Please add `.` at the end for consistency:
```suggestion
      # Enable/disable feature to check constructors are placed after struct declaration.
```

manuelarte replied:
Done!

alexandear commented on .golangci.next.reference.yml:
```diff
@@ -529,6 +531,14 @@ linters:
+      # Enable/disable feature to check constructors are placed after struct declaration
+      # Default: true
+      constructor: false
+      # Enable/disable feature to check whether the exported struct's methods are placed before the non-exported
```
```suggestion
      # Enable/disable feature to check whether the exported struct's methods are placed before the non-exported.
```

manuelarte replied:
Done!

ldez reviewed (approved):
LGTM
